feat(focus): Add arrow support for md-menu and other improvements

milesap opened the pull request:
## Description
- Removed requirement for calling init(), now initialized before
app-seed but after jQuery (required)
- Added new HTML and jQuery element prototype rvFocus() which replaces focus(true) when
setting focus on viewer elements. They are now exclusive and will raise
a warning when used incorrectly.
- Added attribute rv-focus-init which allows the use of focus() on a
viewer element or one of its children if focus is outside the element
- Added jQuery element prototype nextFocus() which moves focus to the
next element
- Documentation added

Closes #1623, #1373

## Testing
Yes, quite a bit

## Documentation
Of course!

## Checklist
<!-- Quick checklist for items that are easy to miss -->

- [x] commits messages follow the guidelines
- [x] code passes unit tests
- [x] release notes have been updated
- [x] PR targets the correct release version

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1741)
<!-- Reviewable:end -->

AleksueiR commented:
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/app/ui/export/custom-size.directive.js, line 21 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-Kehvfp--UJGBYUeVW7Y:-Kehvfp--UJGBYUeVW7Z:b-6oh956) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2c1d42dcae45602542611dafde9a26e158c60d28/src/app/ui/export/custom-size.directive.js#L21)):*
> ```JavaScript
>             restrict: 'E',
>             templateUrl: 'app/ui/export/custom-size.html',
>             link: (scope, el) => el.find('input').first().focus(true)
> ```

Should this be `rvFocus`?

---

*[src/app/ui/export/custom-size.html, line 1 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-Kehtra1UVrJPFAqqgWk:-Kehtra2xzJdsP3vBQIA:b-ym6i52) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2c1d42dcae45602542611dafde9a26e158c60d28/src/app/ui/export/custom-size.html#L1)):*
> ```HTML
> <md-input-container ng-disabled="self.isError">
> ```

Why do you need to split out this template section? There should be a different way to move the focus without splitting this into a separate directive.

The focus is moved to the "width" field when the export dialog itself is opened, not when the custom size sub-panel is opened.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/app/ui/export/custom-size.directive.js, line 21 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-Kehvfp--UJGBYUeVW7Y:-KehxXNADK33Dy9cMr5l:b-3p9coh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2c1d42dcae45602542611dafde9a26e158c60d28/src/app/ui/export/custom-size.directive.js#L21)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

Should this be `rvFocus`?
</blockquote></details>

Missed this one, changed it

---

*[src/app/ui/export/custom-size.html, line 1 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-Kehtra1UVrJPFAqqgWk:-KehxpD225SD1slOjHu2:buy7j4b) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2c1d42dcae45602542611dafde9a26e158c60d28/src/app/ui/export/custom-size.html#L1)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

Why do you need to split out this template section? There should be a different way to move the focus without splitting this into a separate directive.

The focus is moved to the "width" field when the export dialog itself is opened, not when the custom size sub-panel is opened.
</blockquote></details>

There is an `md-content` parent element with `ng-if="self.exportSizes.isCustomOptionSelected()"` so that content is destroyed and re-created each time custom size option is selected in the dropdown. This just sets focus each time it is created (the "angular way" I suppose, instead of watching the `ng-if` for changes and setting focus that way. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
I found few issues but the good point is I am able to navigate with tab in FF and Safari

When we close with esc, the focus didn't go back to export inside the menu panel. Focus only goes to the button to open the panel.
There is a hidden focus between the size selection and the setting button.
When I tab to select custom size and go back to size selector with tab and select Default the focus goes to url bar. It should stay on size selector.

I know it is not part of this PR but if you select Share then press esc, focus goes on the map not the Share button. Same with Basemap, when we close it focus don't go tot he right place. When we change language, panel close and focus goes  to the wrong place and when I tab and I am on menu, it stay white (don't see the selection). A new issue should be created to deal with focus and sidenav panel.

---

Reviewed 16 of 18 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/app/ui/filters/filters-default.directive.js, line 415 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-KeiIOGROihx_4nO7C6_:-KeiIOGROihx_4nO7C6a:btamx4w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/e4b2700c19af65a5fde294d0a4566b8f4cc69b42/src/app/ui/filters/filters-default.directive.js#L415)):*
> ```JavaScript
>                     // at the same time it solve a problem because when focus is on menu button, even if focus is on the
>                     // table cell it goes inside the menu and loop through it at the same time as we navigate the table
>                     $timeout(() => { $rootElement.find('[type=\'filters\'] button.rv-close').rvFocus(); }, 100);
> ```

Whit the rv-close-button, do we still need this? should it go automatically to the close button?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KeiDB5lXXkOtSm-OuZI:bqa2b9e)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
In FF and Chrome when I select something in the menu show/hide all, the focus doesn't stay on the icon to open the menu. It goes at the en of TOC for Safari and it is lost then goes to map with FF. Works in Chrome.

When panel is close, focus doesn't go to export button. In Chrome it goes to menu, in FF and Safari it is lost and the first tab goes to overview map

hidden tab solved!
Change from custom size solved!

---

Review status: 14 of 19 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KejEMO1dTvOdlHXdBV6:b-2mk2ot)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 13 of 18 files at r1, 1 of 2 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741)*
<!-- Sent from Reviewable.io -->

milesap commented:
Thanks for the detailed testing! Please only test in chrome for now, there is no FF or IE support. If any part of FM works in those browsers then bonus points :+1: 

Made some fairly large breaking changes, so a good run through with chrome would be a good idea.

---

Review status: 15 of 21 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/app/ui/filters/filters-default.directive.js, line 415 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-KeiIOGROihx_4nO7C6_:-KejSgXxMDahu5tBuOH3:bks9wx7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/e4b2700c19af65a5fde294d0a4566b8f4cc69b42/src/app/ui/filters/filters-default.directive.js#L415)):*
<details><summary><i>Previously, jolevesq (Johann Levesque) wrote…</i></summary><blockquote>

Whit the rv-close-button, do we still need this? should it go automatically to the close button?
</blockquote></details>

Good point, I haven't had time to test this out but yes it should work (and I'll update this on a subsequent FM update, this one is getting large)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KejTCmkRx-_pyO0ExKw:b-17049c)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
I made a good run an list some issues (only from Chrome this time :smile: ). I don't know if they are all part of this PR but it can help for future WCAG/focus tasks

If I open a menu like collapse all and instead of the arrow to navigate I press tab, the focus goes to url bar and menu can't close. For menu when there is more then one active option you can tab and it will react like press arrow. If there is only one choice, press tab will set focus outside the menu.

when press zoom to location it shows a toast message. The focus should go to the close button of this message instead of having to tab. When we close the message the focus is lost when it should go back to zoom to location.

When we open the help it doesn't focus to close button. When click on close it goes back to help button :thumbsup: 

When press enter on geosearch focus doesn't go to the text input of geosearch. We have to tab once.

Load file or service when I press close button, focus goes back to add layer  :thumbsup: but when I press esc, the TOC is close and focus goes to layer button when it should go on add layer

Datatable seems good. I am able to open from button and panel and when i close from esc or button focus goes at the right place

When open Metadata from the layer setting, focus should go to close button. When we close something open from an item inside a menu I am not sure if focus should go back to the menu item or the button to open the menu is ok?

When we close the export, focus should go back to export button inside sidenav. 

Just a tought, when we tab from map focus it should go to menu instead of overview map. Overview should be focus after  menu panel. This way the focus don't go right/left/right and menu is more important to go to then overview map.

---

Reviewed 2 of 5 files at r3, 5 of 6 files at r4.
Review status: 20 of 21 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KejXcR4UDQ8-6VRB-9J:b71e8yx)*
<!-- Sent from Reviewable.io -->

milesap commented:
> If I open a menu like collapse all and instead of the arrow to navigate I press tab, the focus goes to url bar and menu can't close. For menu when there is more then one active option you can tab and it will react like press arrow. If there is only one choice, press tab will set focus outside the menu.

Fixed

> When we open the help it doesn't focus to close button. When click on close it goes back to help button 👍

Fixed

> When press enter on geosearch focus doesn't go to the text input of geosearch. We have to tab once.

Not fixed in this PR. To implement this there are two options:
- Set focus on the input with rvFocus
- Set input tabindex to -3 and let autocomplete set focus itself
This would need a decorator as the input is not available in `geosearch-bar.js` link function

> Load file or service when I press close button, focus goes back to add layer 👍 but when I press esc, the TOC is close and focus goes to layer button when it should go on add layer

This is correct behaviour for FM since the TOC is closed, layer button is only valid target.

> When open Metadata from the layer setting, focus should go to close button. When we close something open from an item inside a menu I am not sure if focus should go back to the menu item or the button to open the menu is ok?

This is correct behaviour

> When we close the export, focus should go back to export button inside sidenav.

This is correct behaviour for FM since the sidenav is closed on export open, menu button is only valid target.

> Just a tought, when we tab from map focus it should go to menu instead of overview map. Overview should be focus after menu panel. This way the focus don't go right/left/right and menu is more important to go to then overview map.

I like this idea. Perhaps in the next update this will be done, but to keep this PR from touching any more files I'll skip this for now. Overview map is inside `<div class="rv-esri-map" rv-init-map tabindex="0" translate translate-attr-aria-label="geo.aria.mapdescription"></div>` I believe so it would need some poking around to change its ordering.

---

Review status: 18 of 22 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KeskKygwkQg28kXJ2TF:b-1rp117)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
Ok to put geosearch in another issue. It will have to be solve because hidden focus is a WCAG fail.

Maybe it can be part of another issue as well but the esc on export bring back to layer instead of add layer is a little bit confusing. We should discuss on this.

We just miss the auto-focus on close when metadata and settings opens and focus is not trap inside the panel.

This is a good FM progress :thumbsup: 

---

Reviewed 1 of 6 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KfHVdEb6pTYf7q3RMhq:b-rawri0)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KfHXzLi2BbklHnQhGuF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

milesap commented:
ty, opened #1770 for geosearch. The others will be under consideration for future implementation.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1741#-:-KfH__v-1YEMojQ4NUAU:b4njmq)*
<!-- Sent from Reviewable.io -->
